File: backend/app_logging.py
"""Rotating runtime log (flightops_hub.log, next to the exe/config - same
CWD-relative, portable-app reasoning as config_manager.CONFIG_FILE).

Separate from flightops_hub.pyw's own crash_log.txt, which is a
last-resort handler for the narrow window before this module has even
been imported/configured (a startup crash before logging exists at all).
This one runs for the app's whole lifetime and is meant to answer "what
was the user doing right before it broke" for a bug report - not just
catch fatal crashes.
"""
import logging
import logging.handlers
import platform

from .version import APP_VERSION

LOG_FILE = "flightops_hub.log"
LOGGER_NAME = "flightops_hub"

_MAX_BYTES = 2 * 1024 * 1024
_BACKUP_COUNT = 1


def setup():
    """Idempotent - safe to call more than once (e.g. from a test harness),
    won't stack duplicate handlers."""
    logger = logging.getLogger(LOGGER_NAME)
    if logger.handlers:
        return logger

    logger.setLevel(logging.INFO)
    handler = logging.handlers.RotatingFileHandler(
        LOG_FILE, maxBytes=_MAX_BYTES, backupCount=_BACKUP_COUNT, encoding="utf-8"
    )
    handler.setFormatter(logging.Formatter("%(asctime)s %(levelname)s %(name)s: %(message)s"))
    logger.addHandler(handler)
    logger.propagate = False

    logger.info("FlightOps Hub %s starting - %s", APP_VERSION, platform.platform())
    return logger


def get_logger(suffix=None):
    """suffix: dotted sub-name, e.g. get_logger('scenery') -> logger named
    'flightops_hub.scenery' - shares the parent's file handler via normal
    logging hierarchy propagation, no extra setup needed."""
    name = LOGGER_NAME if not suffix else f"{LOGGER_NAME}.{suffix}"
    return logging.getLogger(name)


def read_recent_lines(max_lines=200):
    """Best-effort tail of the current log file, for bundling into a
    diagnostics export - never raises, returns "" if the file can't be read
    (e.g. logging was never set up, or this is a fresh install)."""
    try:
        with open(LOG_FILE, "r", encoding="utf-8", errors="replace") as f:
            lines = f.readlines()
        return "".join(lines[-max_lines:])
    except OSError:
        return ""

File: dotnet/src/FlightOpsHub.Core/AppLogging.cs
using System.Text;

namespace FlightOpsHub.Core;

/// <summary>
/// Port of backend/app_logging.py - a small rotating runtime log file, next
/// to the exe/config (same CWD-relative, portable-app reasoning as
/// ConfigManager.ConfigFileName). Minimal hand-rolled rotation rather than
/// a logging library dependency, matching the size of what Python's own
/// RotatingFileHandler usage here actually needed.
/// </summary>
public static class AppLogging
{
    public const string LogFileName = "flightops_hub.log";

    private const long MaxBytes = 2 * 1024 * 1024;

    private static readonly object Lock = new();

    public static void Info(string path, string message) => Write(path, "INFO", message);

    public static void Warning(string path, string message) => Write(path, "WARNING", message);

    public static void Error(string path, string message) => Write(path, "ERROR", message);

    private static void Write(string path, string level, string message)
    {
        lock (Lock)
        {
            try
            {
                RotateIfNeeded(path);
                var line = $"{DateTime.Now:yyyy-MM-dd HH:mm:ss,fff} {level} flightops_hub: {message}{Environment.NewLine}";
                File.AppendAllText(path, line, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false));
            }
            catch
            {
                // Logging must never be the reason the app crashes.
            }
        }
    }

    private static void RotateIfNeeded(string path)
    {
        if (!File.Exists(path)) return;
        if (new FileInfo(path).Length < MaxBytes) return;

        var backupPath = path + ".1";
        try
        {
            File.Delete(backupPath);
        }
        catch
        {
            // Nothing to clean up.
        }
        File.Move(path, backupPath);
    }

    /// <summary>
    /// Best-effort tail of the current log file, for bundling into a
    /// diagnostics export - never throws, returns "" if the file can't be
    /// read (e.g. logging was never set up, or this is a fresh install).
    /// </summary>
    public static string ReadRecentLines(string path, int maxLines = 200)
    {
        try
        {
            var lines = File.ReadAllLines(path);
            var start = Math.Max(0, lines.Length - maxLines);
            return string.Join(Environment.NewLine, lines[start..]) + (lines.Length > 0 ? Environment.NewLine : "");
        }
        catch
        {
            return "";
        }
    }
}
